trillian/ctfe/handlers.go

```diff
@@ -568,9 +573,17 @@ func getProofByHash(ctx context.Context, c *LogContext, w http.ResponseWriter, r
 		return c.toHTTPStatus(err), fmt.Errorf("backend GetInclusionProofByHash request failed: %v", err)
 	}
 
+	// We could fail to get the proof because the tree size that the server has
+	// is not large enough.
+	if rsp.GetSignedLogRoot() != nil && rsp.GetSignedLogRoot().TreeSize < treeSize {
+		return http.StatusNotFound, fmt.Errorf("log returned tree size: %d but we expected: %d", rsp.SignedLogRoot.TreeSize, treeSize)
+	}
+
 	// Additional sanity checks
 	if len(rsp.Proof) == 0 {
-		return http.StatusInternalServerError, errors.New("get-proof-by-hash: backend did not return a proof")
+		// This is no longer a 5xx error as backend returns the STH. Previously
```

Comment:
Comments that talk about how the code used to be go out of date quickly -- can we just describe what's happening/why?  E.g. "The backend returns the STH even when there is no proof, so explicitly map this to 4xx"

Reply:
Done.